tensorboard/data/server/logdir.rs

```diff
@@ -15,68 +15,48 @@ limitations under the License.
+use std::collections::HashMap;
+use std::io::{self, Read};
 use std::path::{Path, PathBuf};
-use walkdir::WalkDir;
 
 use crate::commit::Commit;
 use crate::run::RunLoader;
 use crate::types::Run;
 
-/// A loader for a log directory, connecting a filesystem to a [`Commit`] via [`RunLoader`]s.
-pub struct LogdirLoader<'a> {
-    commit: &'a Commit,
-    logdir: PathBuf,
-    runs: HashMap<Run, RunState>,
-    checksum: bool,
-}
+/// A TensorBoard log directory, with event files organized into runs.
+pub trait Logdir {
+    /// Type of output stream for reading event files under this log directory.
+    type File: Read;
 
-struct RunState {
-    /// Stateful loader for this run.
-    loader: RunLoader,
-    /// Path to this run's directory relative to the root logdir.
-    relpath: PathBuf,
-    /// Logdir-relative paths to other directories that normalize to the same name as this run
-    /// after lossy Unicode conversion.
-    ///
-    /// Directories on the filesystem need not be valid Unicode. On Unix, for instance, they're
-    /// arbitrary byte sequences without `\0` or `/`. But TensorBoard wants run names to be strings
-    /// for human display. Thus, we apply [`OsStr::to_string_lossy`] to convert run directory paths
-    /// to run names. As the name suggests, this is a lossy conversion: it replaces non-Unicode
-    /// sequences with U+FFFD. This creates an edge case wherein two different directories may map
-    /// to the same run name. In such a case, we merge the two directories into one run, and print
-    /// a warning.
+    /// Finds all event files under the log directory.
     ///
-    /// This set tracks the directories other than `self.relpath` that comprise this run. It is
-    /// non-empty only when we hit this edge case: i.e., when there are multiple filesystem paths
-    /// that are invalid Unicode and collide. It's used for deduplicating warnings, so that we
-    /// don't warn about the directory again on every load cycle.
+    /// Event files within each run should be emitted in chronological order. Canonically, a file
+    /// is an event file if its basename contains [`EVENT_FILE_BASENAME_INFIX`] as a substring.
+    fn discover(&self) -> io::Result<HashMap<Run, Vec<PathBuf>>>;
+
+    /// Attempts to open an event file for reading.
     ///
-    /// [`OsStr::to_string_lossy`]: std::ffi::OsStr::to_string_lossy
-    collided_relpaths: HashSet<PathBuf>,
+    /// The `path` should be one of the values returned by a previous call to [`Self::discover`].
+    fn open(&self, path: &Path) -> io::Result<Self::File>;
 }
 
-/// Record of an event file found under the log directory.
-#[derive(Debug, PartialEq, Eq)]
-struct EventFileDiscovery {
-    run_relpath: PathBuf,
-    event_file: PathBuf,
+/// A file is treated as an event file if its basename contains this substring.
+pub const EVENT_FILE_BASENAME_INFIX: &str = "tfevents";
+
+/// A loader for a log directory, connecting a filesystem to a [`Commit`] via [`RunLoader`]s.
```

Comment:
Maybe worth expanding this comment or adding field-level comments?  Now that `logdir` is a more advanced entity, would be nice to outline in slightly more detail how it relates to the other fields.

Reply:
Sure, can do.